Add a DR overview focused on resiliency with comparison for HA & DR (#18490)
cockroachdb/docs: src/current/v24.1/disaster-recovery-overview.md

---
title: Disaster Recovery Overview
summary: Learn how to make your cluster resilient with high availability and disaster recovery features.
toc: true
---

> Resilient deployments aim for continuity in database operation to protect from data loss and down time. To maintain resiliency, it is necessary to build deployments with _high availability_ and _disaster recovery_ coverage.

Reviewer: database operation continuity?

Author: I haven't changed this. Having "database operation" modify "continuity" feels a little harder to read. I have left as-is for now — hopefully my docs review partner may have an idea here.

Reviewer: No worries, sounds good!

> Resilient deployments aim for continuity in database operation to protect from data loss and down time. To maintain resiliency, it is necessary to build deployments with _high availability_ and _disaster recovery_ coverage.

Reviewer 2: ```suggestion
Resilient deployments aim for continuity in database operations to protect from data loss and downtime. To maintain resiliency, it is necessary to build deployments with _high availability_ and _disaster recovery_ coverage.
```

- [High availability](#choose-a-high-availability-strategy): Continuous and uninterrupted access to data even in the presence of failures or disruptions to maximize uptime.

> - [Disaster recovery](#choose-a-disaster-recovery-strategy): Recovery from a major incident or disaster to minimize downtime and data loss.

Author: Yes, like it. To match the other bullet point, I have changed so they are both verbs.

To build resilient cluster deployments, CockroachDB provides both [built-in replication]({% link {{ page.version.version }}/architecture/replication-layer.md %}) for high availability and disaster recovery features for coverage during unplanned incidents.

{{site.data.alerts.callout_success}}
For a practical guide on how CockroachDB replicates, distributes, and rebalances data, refer to the [Replication and Rebalancing]({% link {{ page.version.version }}/demo-replication-and-rebalancing.md %}) demo.
{{site.data.alerts.end}}

## Resilience strategy

As you evaluate CockroachDB's disaster recovery features, consider your organization's requirements for the amount of tolerable data loss and the acceptable length of time to recover.

> - Recovery Point Objective (RPO): The maximum amount of time that an organization can tolerate losing data.

Reviewer: "maximum amount of data loss – as measured by time – that an organization can tolerate." Maybe this?

Author: Added, with parenthesis.

> - Recovery Point Objective (RPO): The maximum amount of data loss (measured by time) that an organization can tolerate losing data.

Reviewer: I think "losing data" should be removed?

- Recovery Time Objective (RTO): The maximum length of time it should take to restore normal operations following an outage.

For example, when you use backups:

<img src="{{ 'images/v24.1/rpo-rto.png' | relative_url }}" alt="Simulating RPO and RTO. With RPO representing the tolerable data loss, and RTO representing the tolerable time to recovery." style="border:0px solid #eee;max-width:100%" />

For a comparison of CockroachDB resiliency features, refer to the following sections:

- [Choose a high availability strategy](#choose-a-high-availability-strategy)
- [Choose a disaster recovery strategy](#choose-a-disaster-recovery-strategy)

### Choose a high availability strategy

> CockroachDB uses synchronous, built-in replication to create and distribute copies of data, ensuring consistency across the data copies. The database can tolerate nodes going offline without service interruption whether in a single-region or [multi-region]({% link {{ page.version.version }}/multiregion-overview.md %}) cluster.

Reviewer 2: ```suggestion
CockroachDB uses synchronous, built-in replication to create and distribute copies of data, ensuring consistency across the data copies. The database can tolerate nodes going offline without service interruption, whether in a single-region or [multi-region]({% link {{ page.version.version }}/multiregion-overview.md %}) cluster.
```

<table class="comparison-chart">
  <tr>
    <th></th>

>     <th>Single-region replication</th>

Reviewer: Is it worth adding "synchronous" replication here?

Author: Added to both columns here.

  </tr>

  <tr>
    <td class="comparison-chart__feature">
      <b>RTO</b>
    </td>
    <td>Zero RTO<br>Potential increased latency for 1-9 seconds</td>
    <td>Zero RTO<br>Potential increased latency for 1-9 seconds</td>
  </tr>

  <tr>
    <td class="comparison-chart__feature">
      <b>Write latency</b>
    </td>
    <td>Region-local write latency<br>p50 latency < 5ms (for multiple availability zones in us-east1)</td>
    <td>Cross-region write latency<br>p50 latency > 50ms  (for a multi-region cluster in us-east1, us-east-2, us-west-1)</td>
  </tr>

  <tr>
    <td class="comparison-chart__feature">
      <b>Recovery</b>
    </td>
    <td>Automatic</td>
    <td>Automatic</td>
  </tr>

  <tr>
    <td class="comparison-chart__feature">
      <b>Hardware cost</b>
    </td>
    <td>1x<br>Networking between availability zones</td>
    <td>1x<br>Networking between datacenters</td>
  </tr>

  <tr>
    <td class="comparison-chart__feature">
      <b>Minimum regions</b>
    </td>
    <td>1</td>
    <td>3</td>
  </tr>

  <tr>
    <td class="comparison-chart__feature">
      <b>Fault tolerance</b>
    </td>
    <td>Zero RPO node, availability zone failures</td>

>     <td>Zero RPO node, availability zone failures</td>

Reviewer: The multi-region one  should also be able to survive a region failure

Author: Done!

  </tr>

</table>

> For detail on designing your cluster topology for high availability with replication, refer to the [Disaster Recovery Planning]({% link {{ page.version.version }}/disaster-recovery-planning.md %}#hardware-failure) page.

Reviewer 2: ```suggestion
For details on designing your cluster topology for high availability with replication, refer to the [Disaster Recovery Planning]({% link {{ page.version.version }}/disaster-recovery-planning.md %}#hardware-failure) page.
```

### Choose a disaster recovery strategy

> CockroachDB is designed to recover automatically, however building [backups]({% link {{ page.version.version }}/backup-and-restore-overview.md %}) or [physical cluster replication]({% link {{ page.version.version }}/physical-cluster-replication-overview.md %}) into your disaster recovery planning is an important part of a resilient deployment.

Reviewer 2: ```suggestion
CockroachDB is designed to recover automatically; however, building [backups]({% link {{ page.version.version }}/backup-and-restore-overview.md %}) or [physical cluster replication]({% link {{ page.version.version }}/physical-cluster-replication-overview.md %}) into your disaster recovery planning is an important part of a resilient deployment.
```

    <th>Point in time backup & restore</th>
    <th>Physical cluster replication (async)</th>
  </tr>

  <tr>
    <td class="comparison-chart__feature">
      <b>RPO</b>
    </td>
    <td>>=5 minutes</td>
    <td>10s of seconds</a></td>
  </tr>

  <tr>
    <td class="comparison-chart__feature">
      <b>RTO</b>
    </td>
    <td>Minutes to hours, depending on data size and number of nodes</td>
    <td>Seconds to minutes, depending on cluster size</td>
  </tr>

  <tr>
    <td class="comparison-chart__feature">
      <b>Write latency</b>
    </td>
    <td>No impact</td>
    <td>No impact</td>
  </tr>

  <tr>
    <td class="comparison-chart__feature">
      <b>Recovery</b>
    </td>
    <td>Manual restore</td>
    <td>Manual failover</td>
  </tr>

  <tr>
    <td class="comparison-chart__feature">
      <b>Hardware cost</b>
    </td>
    <td>1x + S3-like storage & networking</td>
    <td>2x + networking between datacenters</td>
  </tr>

  <tr>
    <td class="comparison-chart__feature">
      <b>Minimum regions</b>
    </td>
    <td>1</td>
    <td>2</td>
  </tr>

  <tr>
    <td class="comparison-chart__feature">
      <b>Fault tolerance</b>
    </td>
    <td>Not applicable</td>

>     <td>Zero RPO node, availability zone region failure with loss up to RPO</td>

Reviewer: nit: i think a comma is needed after 'availability zone'?

Author: Oh yeah, looks like it! Added the comma!
